File: reactor/java-reactor/src/main/java/com/example/reactor/javareactor/service/UserService.java
package com.example.reactor.javareactor.service;

import com.example.reactor.javareactor.dto.request.CreateUserRequest;
import com.example.reactor.javareactor.dto.response.UserResponse;
import com.example.reactor.javareactor.entity.User;
import com.example.reactor.javareactor.exception.UserAlreadyExistException;
import com.example.reactor.javareactor.exception.UserNotFoundException;
import com.example.reactor.javareactor.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Mono;

@RequiredArgsConstructor
@Service
public class UserService {
    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public Mono<UserResponse> getUserById(final String id) {
        return userRepository.findById(id)
                             .switchIfEmpty(Mono.error(new UserNotFoundException()))
                             .map(UserResponse::from);
    }

    public Mono<UserResponse> createUser(final CreateUserRequest request) {
        return userRepository.findByEmail(request.email())
                             .flatMap(user -> Mono.<User>error(new UserAlreadyExistException()))
                             .switchIfEmpty(
                                 Mono.defer(
                                     () -> userRepository.save(
                                         User.builder()
                                             .email(request.email())
                                             .password(passwordEncoder.encode(request.password()))
                                             .name(request.name())
                                             .build()
                                     )
                                 )
                             )
                             .map(UserResponse::from);
    }
}

File: reactor/kotlin-reactor/src/main/kotlin/com/example/reactor/kotlinreactor/service/UserService.kt
package com.example.reactor.kotlinreactor.service

import com.example.reactor.kotlinreactor.dto.request.CreateUserRequest
import com.example.reactor.kotlinreactor.dto.response.UserResponse
import com.example.reactor.kotlinreactor.entity.User
import com.example.reactor.kotlinreactor.exception.UserAlreadyExistException
import com.example.reactor.kotlinreactor.exception.UserNotFoundException
import com.example.reactor.kotlinreactor.repository.UserRepository
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class UserService(
    private val userRepository: UserRepository,
    private val passwordEncoder: PasswordEncoder
) {
    fun getUserById(id: String): Mono<UserResponse> =
        userRepository.findById(id)
            .switchIfEmpty(Mono.error(UserNotFoundException()))
            .map { UserResponse(it) }

    fun createUser(request: CreateUserRequest): Mono<UserResponse> =
        with(request) {
            userRepository.findByEmail(email)
                .flatMap { Mono.error<User>(UserAlreadyExistException()) }
                .switchIfEmpty(
                    Mono.defer {
                        userRepository.save(
                            User(
                                email = email,
                                password = passwordEncoder.encode(password),
                                name = name
                            )
                        )
                    }
                )
                .map { UserResponse(it) }
        }
}
